core/src/transformers/providers/voyage.rs

```diff
@@ -0,0 +1,152 @@
+        VoyageEmbeddingBody {
+            input: request.input,
+            model: request.model,
+            input_type: "document".to_string(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct VoyageEmbeddingResponse {
+    pub data: Vec<EmbeddingObject>,
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct EmbeddingObject {
+    pub embedding: Vec<f64>,
+}
+
+impl From<VoyageEmbeddingResponse> for GenericEmbeddingResponse {
+    fn from(response: VoyageEmbeddingResponse) -> Self {
+        GenericEmbeddingResponse {
+            embeddings: response.data.iter().map(|x| x.embedding.clone()).collect(),
+        }
+    }
+}
+
+impl VoyageProvider {
+    pub fn new(url: Option<String>, api_key: Option<String>) -> Self {
+        let final_url = match url {
+            Some(url) => url,
+            None => VOYAGE_BASE_URL.to_string(),
+        };
+        let final_api_key = match api_key {
+            Some(api_key) => api_key,
+            None => env::var("VOYAGE_API_KEY").expect("VOYAGE_API_KEY not set"),
+        };
+        VoyageProvider {
+            url: final_url,
+            api_key: final_api_key,
+        }
+    }
+}
+
+#[async_trait]
+impl EmbeddingProvider for VoyageProvider {
+    async fn generate_embedding<'a>(
+        &self,
+        request: &'a GenericEmbeddingRequest,
+    ) -> Result<GenericEmbeddingResponse, VectorizeError> {
+        let client = Client::new();
+
+        let req_body = VoyageEmbeddingBody::from(request.clone());
+        let embedding_url = format!("{}/embeddings", self.url);
+
+        let response = client
+            .post(&embedding_url)
+            .timeout(std::time::Duration::from_secs(120_u64))
+            .header("Content-Type", "application/json")
+            .header("Authorization", format!("Bearer {}", self.api_key))
+            .json(&req_body)
+            .send()
+            .await?;
+
+        let embeddings = handle_response::<VoyageEmbeddingResponse>(response, "embeddings").await?;
+        Ok(GenericEmbeddingResponse {
+            embeddings: embeddings
+                .data
+                .iter()
+                .map(|x| x.embedding.clone())
+                .collect(),
+        })
+    }
+
+    async fn model_dim(&self, model_name: &str) -> Result<u32, VectorizeError> {
+        Ok(voyager_embedding_dim(model_name) as u32)
+    }
+}
+
+pub fn voyager_embedding_dim(model_name: &str) -> i32 {
+    match model_name {
+        "voyage-3-lite" => 512,
+        "voyage-3" | "voyage-finance-2" | "voyage-multilingual-2" | "voyage-law-2" => 1024,
+        "voyage-code-2" => 1536,
+        // older models
+        "voyage-large-2" => 1536,
+        "voyage-large-2-instruct"
+        | "voyage-2"
+        | "voyage-lite-02-instruct"
+        | "voyage-02"
+        | "voyage-01"
+        | "voyage-lite-01"
+        | "voyage-lite-01-instruct" => 1024,
+        _ => 1536,
+    }
+}
```

Comment:
I was thinking about this, I wonder if rather than specifying for each model, we could just call the model with a single token and then get the length of the array that is returned.